### DISCUSSION TITLE
on first call API resposnds error 500 after 15 minutes of inactivity

### DISCUSSION THREAD

**Discussion by sandros94** [directus/directus#17663] [Archive]
This discussion is a continuation of directus-community/nuxt-directus#107
## Summary
After 15 minutes of idle time the very first API call responds `{"errors":[{"message":"An unexpected error occurred.","extensions":{"code":"INTERNAL_SERVER_ERROR"}}]}`.
The content is requested behind a static access token from a manually created API user (no email) that has Read Only permission to most of the content in the collections.
Immediately refreshing/resubmitting the request responds 200 with the requested data.

## Relevant logs
The most recent finding is:
```
[00:52:28.512] INFO: Server started at http://0.0.0.0:8055
[01:35:03] GET /items/website?access_token=--redacted-- 500 17ms
[01:35:03] GET /favicon.ico 500 8ms
[01:35:03.727] ERROR: select "directus_users"."id", "directus_users"."role", "directus_roles"."admin_access", "directus_roles"."app_access" from "directus_users" left join "directus_roles" on "directus_users"."role" = "directus_roles"."id" where "directus_users"."token" = $1 and "status" = $2 limit $3 - Connection terminated unexpectedly
    err: {
      "type": "Error",
      "message": "select \"directus_users\".\"id\", \"directus_users\".\"role\", \"directus_roles\".\"admin_access\", \"directus_roles\".\"app_access\" from \"directus_users\" left join \"directus_roles\" on \"directus_users\".\"role\" = \"directus_roles\".\"id\" where \"directus_users\".\"token\" = $1 and \"status\" = $2 limit $3 - Connection terminated unexpectedly",
      "stack":
          Error: select "directus_users"."id", "directus_users"."role", "directus_roles"."admin_access", "directus_roles"."app_access" from "directus_users" left join "directus_roles" on "directus_users"."role" = "directus_roles"."id" where "directus_users"."token" = $1 and "status" = $2 limit $3 - Connection terminated unexpectedly
              at Connection.<anonymous> (/directus/node_modules/pg/lib/client.js:131:73)
              at Object.onceWrapper (node:events:627:28)
              at Connection.emit (node:events:513:28)
              at Socket.<anonymous> (/directus/node_modules/pg/lib/connection.js:62:12)
              at Socket.emit (node:events:513:28)
              at TCP.<anonymous> (node:net:322:12)
    }
[01:35:03.834] WARN: Connection Error: Connection ended unexpectedly
[01:35:03.839] ERROR: select "ip_access" from "directus_roles" where "id" is null limit $1 - Connection terminated unexpectedly
    err: {
      "type": "Error",
      "message": "select \"ip_access\" from \"directus_roles\" where \"id\" is null limit $1 - Connection terminated unexpectedly",
      "stack":
          Error: select "ip_access" from "directus_roles" where "id" is null limit $1 - Connection terminated unexpectedly
              at Connection.<anonymous> (/directus/node_modules/pg/lib/client.js:131:73)
              at Object.onceWrapper (node:events:627:28)
              at Connection.emit (node:events:513:28)
              at Socket.<anonymous> (/directus/node_modules/pg/lib/connection.js:62:12)
              at Socket.emit (node:events:513:28)
              at TCP.<anonymous> (node:net:322:12)
    }
```
> Connection terminated unexpectedly

As pointed out to me, this could mean a connection drop between directus and the database.
The error in the log doesn't show on every 500, but still on most of them. More importantly it will always show on the first 500 after the server has been deployed.

## At what point I'm at
I first noticed this problem at the beginning of January, with the first directus deployment mid of Dec (but I wasn't using the static access token iirc).

While testing **many** times I've noticed that 15 minutes is a good minimum to trigger an error 500. Looking at directus docs the only related thing to 15m was `ACCESS_TOKEN_TTL` but changing it to 1m/20m didn't change the behaviour of the error.

### Talking about the log above
Directus is currently self hosted on a OVH vps. On this very vps I've already deployed in the past a number of databases, one of witch was a postgresql that had to deal with time-sensitive data. So I'm not fully convinced that it's fully related to vps/hardware standby (or similar features).
But surely this is my first deployment that involves redis cache, something that I've always looked into and wanted to learn to accelerate some databases while scaling horizontally (currently only as a personal project). 

## My deployment
The docker-compose I'm using to deploy directus on a single-node swarm (created following [directus doc](https://docs.directus.io/self-hosted/docker-guide.html#docker-compose)):
``` yml
version: '3.4'

services:
  frontend:
    image: ${IMG_REPO}
    
    networks:
      - traefik-public

    stdin_open: true
    tty: true

    deploy:
      mode: replicated
      replicas: 1
      placement:
        constraints: 
          - ${NODE_PLACEMENT}
      labels:
        - "traefik.enable=true"
        - "traefik.http.routers.${PROJECT_NAME}-frontend.rule=Host(`${FRONTEND_DOMAIN}`)"
        - "traefik.http.routers.${PROJECT_NAME}-frontend.tls.certresolver=letsencrypt"
        - "traefik.http.routers.${PROJECT_NAME}-frontend.entrypoints=websecure"
        - "traefik.http.routers.${PROJECT_NAME}-frontend.service=${PROJECT_NAME}-frontendsvc"
        - "traefik.http.services.${PROJECT_NAME}-frontendsvc.loadbalancer.server.port=3000"


  directus:
    image: directus/directus:9

    volumes:
      # currently binding for practical reasons
      - ${LOCAL_PATH}/directus-uploads:/directus/uploads
      #- directus-uploads:/directus/uploads
      #- directus-extensions:/directus/extensions
    networks:
      - directus-network
      - traefik-public

    tty: true
    stdin_open: true
    environment:
      KEY: '${UNIQUE_KEY}'
      SECRET: '${UNIQUE_SECRET}'
      DB_CLIENT: 'pg'
      DB_HOST: 'database'
      DB_PORT: '5432'
      DB_DATABASE: '${PROJECT_NAME}-db'
      DB_USER: '${DB_USER}'
      DB_PASSWORD: '${DB_PASSWORD}'
      DB_POOL_MIN: '0'
      CACHE_ENABLED: 'true'
      CACHE_STORE: 'redis'
      CACHE_REDIS: 'redis://cache:6379'
      CACHE_AUTO_PURGE: 'true'
      ADMIN_EMAIL: '${ADMIN_EMAIL}'
      ADMIN_PASSWORD: '${ADMIN_PASSWORD}'
      PUBLIC_URL: 'https://${DIRECTUS_URL}'
      ASSETS_TRANSFORM_IMAGE_MAX_DIMENSION: '8256'
      # testing stuff with CORS
      CORS_ENABLED: "true"
      CORS_ORIGIN: "true"
    deploy:
      restart_policy:
        condition: on-failure
        delay: 3s
      mode: replicated
      replicas: 1
      placement:
        constraints: 
          - ${NODE_PLACEMENT}
      labels:
        - "traefik.enable=true"
        - "traefik.http.routers.${PROJECT_NAME}-backend.rule=Host(`${DIRECTUS_URL}`)"
        - "traefik.http.routers.${PROJECT_NAME}-backend.tls.certresolver=letsencrypt"
        - "traefik.http.routers.${PROJECT_NAME}-backend.entrypoints=websecure"
        - "traefik.http.routers.${PROJECT_NAME}-backend.service=${PROJECT_NAME}-backendsvc"
        - "traefik.http.services.${PROJECT_NAME}-backendsvc.loadbalancer.server.port=8055"


  database:
    image: postgis/postgis:13-master
    tty: true
    stdin_open: true
    volumes:
      - directus-db:/var/lib/postgresql/data
      # quick pg backup folder
      - ${LOCAL_PATH}/pg_dumps:/pg_dumps
    networks:
      - directus-network
    environment:
      POSTGRES_DB: '${PROJECT_NAME}-db'
      POSTGRES_USER: '${DB_USER}'
      POSTGRES_PASSWORD: '${DB_PASSWORD}'


  cache:
    image: redis:6
    networks:
      - directus-network


volumes:
  directus-uploads:
  #directus-extensions:
  directus-db:


networks:
  directus-network:
  traefik-public:
    external: true
```

## Why is this important to me?
At this point I'm 99% sure it's a miss-configuration on my side, not a directus problem, but all of this is mainly to **learn** how to track down a problem and possibly solve it.
Thanks to intevel and rijkvanzanten (intentionally not tagging them to not trigger a useless notification), that followed along while I was learning.

**Comment by sandros94** [comment]
While doing some more research I started thinking: could it be that, since I'm not explicitly managing any header forwarding via Traefik, that causes Directus to fail as the log above?

> ```
> [00:52:28.512] INFO: Server started at http://0.0.0.0:8055
> [01:35:03] GET /items/website?access_token=--redacted-- 500 17ms
> [01:35:03] GET /favicon.ico 500 8ms
> [01:35:03.727] ERROR: select "directus_users"."id", "directus_users"."role", "directus_roles"."admin_access", "directus_roles"."app_access" from "directus_users" left join "directus_roles" on "directus_users"."role" = "directus_roles"."id" where "directus_users"."token" = $1 and "status" = $2 limit $3 - Connection terminated unexpectedly
>     err: {
>       "type": "Error",
>       "message": "select \"directus_users\".\"id\", \"directus_users\".\"role\", \"directus_roles\".\"admin_access\", \"directus_roles\".\"app_access\" from \"directus_users\" left join \"directus_roles\" on \"directus_users\".\"role\" = \"directus_roles\".\"id\" where \"directus_users\".\"token\" = $1 and \"status\" = $2 limit $3 - Connection terminated unexpectedly",
>       "stack":
>           Error: select "directus_users"."id", "directus_users"."role", "directus_roles"."admin_access", "directus_roles"."app_access" from "directus_users" left join "directus_roles" on "directus_users"."role" = "directus_roles"."id" where "directus_users"."token" = $1 and "status" = $2 limit $3 - Connection terminated unexpectedly
>               at Connection.<anonymous> (/directus/node_modules/pg/lib/client.js:131:73)
>               at Object.onceWrapper (node:events:627:28)
>               at Connection.emit (node:events:513:28)
>               at Socket.<anonymous> (/directus/node_modules/pg/lib/connection.js:62:12)
>               at Socket.emit (node:events:513:28)
>               at TCP.<anonymous> (node:net:322:12)
>     }
> [01:35:03.834] WARN: Connection Error: Connection ended unexpectedly
> [01:35:03.839] ERROR: select "ip_access" from "directus_roles" where "id" is null limit $1 - Connection terminated unexpectedly
>     err: {
>       "type": "Error",
>       "message": "select \"ip_access\" from \"directus_roles\" where \"id\" is null limit $1 - Connection terminated unexpectedly",
>       "stack":
>           Error: select "ip_access" from "directus_roles" where "id" is null limit $1 - Connection terminated unexpectedly
>               at Connection.<anonymous> (/directus/node_modules/pg/lib/client.js:131:73)
>               at Object.onceWrapper (node:events:627:28)
>               at Connection.emit (node:events:513:28)
>               at Socket.<anonymous> (/directus/node_modules/pg/lib/connection.js:62:12)
>               at Socket.emit (node:events:513:28)
>               at TCP.<anonymous> (node:net:322:12)
>     }
> ```

My current Traefik deployment is a fork of [this one](https://github.com/digitoolmedia/dm-docker-potrec/blob/stable/swarm-potrec.yml)

**Reply by sandros94** [reply]
Ok by default Traefik forwards [some headers](https://doc.traefik.io/traefik/getting-started/faq/#what-are-the-forwarded-headers-when-proxying-http-requests).

**Reply by sandros94** [reply]
Yes, my headers were configured and working correctly for https, but didn't get redirected useful http ones. Even more I wasn't forwarding correctly my XFF, now it's fixed too.

At this point I need to understand why it's `directus_user` collection the error, why not `website` collection?

**Comment by sandros94** [comment]
I started digging it quite deep:
Looks like, as @rijkvanzanten already suggested, that pg will drop the connection. ~~More precisely something along the line that on slow CPU/IO storage (or virtualized one like the vps I'm using) is leading Knex to report `Connection terminated unexpectedly` after an idle time. My undestanding is that after x amount of time (between 15 and 20 mins) the pg data is read from storage instead of memory.~~ Looks like it's the default behavior of Docker Swarm dropping idle conntections on overlay netwroks, look the link in the reply below.

There are a few issue about this on Knex repo ([one of which](https://github.com/knex/knex/issues/3523) I'm studying).
If I read [Directus docs](https://docs.directus.io/self-hosted/config-options.html#database) correctly I should be able to pass variables to Knex using `DB_POOL__*`:

> Pooling
> 
> All the DB_POOL__ prefixed options are passed to [tarn.js](https://github.com/vincit/tarn.js#usage) through [Knex](http://knexjs.org/#Installation-pooling)

As soon as I understand what variable I need to pass to configure the timeout I'll update.

**Reply by sandros94** [reply]
Oh, I didn't remember I've already read #8634 and passed `DB_POOL_MIN=0`.
But didn't test `DB_POOL_MIN=number:0` as explained [here](https://github.com/directus/directus/issues/4214#issuecomment-785946066) and [here](https://docs.directus.io/self-hosted/config-options.html#environment-syntax-prefix)

But I'm almost sure I'm on to something [related to Docker Swarm in general](https://github.com/moby/moby/issues/31208)

**Reply by sandros94** [reply]
I see that in the Directus docs (liked above) about `DB_POOL__*` have two underscores, while in the past issues in this repo are referred as `DB_POOL_*` with a single underscore. Where should I see this code and how could I verify it?

**Reply by sandros94** [reply]
What made me think it could be related to Docker Swarm was reading the pg log:
```
[...]
2023-03-03 12:15:28.523 UTC [22633] LOG:  could not receive data from client: Connection reset by peer
2023-03-03 12:15:28.523 UTC [22711] LOG:  could not receive data from client: Connection reset by peer
2023-03-03 12:43:52.459 UTC [22785] LOG:  could not receive data from client: Connection reset by peer
2023-03-03 13:03:32.107 UTC [22786] LOG:  could not receive data from client: Connection reset by peer
2023-03-03 13:27:33.899 UTC [23015] LOG:  could not receive data from client: Connection reset by peer
2023-03-03 13:31:23.275 UTC [22997] LOG:  could not receive data from client: Connection reset by peer
[...]
```

So I started digging into it.
So far the changes I've made seems successful, and I see no more log errors both in directus and pg for the past 22h. But I want to test them a bit more to see if all are needed or there was a key one. The two most important ones are inside docker-compose, one related to directus (Knex `DB_POOL_MIN`) and one related to the pg deployment (using the [routing mode dnsrr for the overlay network](https://docs.docker.com/network/overlay/#bypass-the-routing-mesh-for-a-swarm-service)).

For visuals, here's the docker-compose changes:
``` yml
version: '3.4'

services:
  directus:
    image: directus/directus:9
    [...]
    environment:
      [...]
      DB_POOL_MIN: 'number:0'
      [...]
    [...]

  database:
    image: postgis/postgis:13-master
    [...]
    deploy:
      endpoint_mode: dnsrr
      restart_policy:
        condition: on-failure
        delay: 3s
      mode: replicated
      replicas: 1
      placement:
        constraints: 
          - ${NODE_PLACEMENT}
```

Additional note: I did forget to add to the pg a deployment constrain, since at the moment I'm using the node storage.

I'm confidently sure that this was it. But again, I would like to test it a bit more before creating a detailed (and condensed) response and marking it as the solution.

**Comment by sandros94** [comment]
## TL;DR
It was indeed `DB_POOL__MIN:0` with **two underscores**, no need to setup `endpoint_mode: dnsrr`. And yes, the problem is caused by Docker Swarm that after 15 mins of inactivity it drops all tcp connections between directus and postgresql.

## What actually happened
### The first take
As I said earlier in one of my reply there have been a couple of issue ([here](https://github.com/directus/directus/issues/8634#issuecomment-938443855) and [here](https://github.com/directus/directus/issues/4214#issuecomment-785946066)) referring to `DB_POOL_MIN` with **single undescores**, and at first (in Directus 9) they  didn't seem to work. So I started to look for the solution on the other side.

### Second take
While reading Docker Swarm related issues (mainly moby/moby#31208) I've discovered the usage of [endpoint dnsrr](https://github.com/moby/moby/issues/31208#issuecomment-397621262), and while I agreed with teh fact that the pg database wouldn't be a Swarm service anyway, but still, if I wanted to use a load balancer in the future I had to config a specific one to work with it (as described [in the docs](https://docs.docker.com/network/overlay/#bypass-the-routing-mesh-for-a-swarm-service)).

> You must run your own load balancer in front of the service. A DNS query for the service name on the Docker host returns a list of IP addresses for the nodes running the service. Configure your load balancer to consume this list and balance the traffic across the nodes

### Third and final ~~dawn~~ take
Not satisfied with the resaults, and the fact that to test it I had to wait 15 minutes each time (🙄) I decided to deploy a bunch of stacks each with its directus, redis and pg services, but slightly changing them in the configuration. Couple of which involving changing how many underscores there were in `DB_POOL_MIN` since while re-reading the docs I've noticed that [it mentioned it](https://docs.directus.io/self-hosted/config-options.html#database) with two underscorse... but it looked strange since the people in the issues above said that after applying them, with what looked like single underscores, worked for them.

But still the docs said:
> Pooling
> 
> All the DB_POOL__ prefixed options are passed to [tarn.js](https://github.com/vincit/tarn.js#usage) through [Knex](http://knexjs.org/#Installation-pooling)

And of course, `DB_POOL__MIN:0` worked first try, and for the past few days I haven't seen in the logs any 500s nor the database log mentioned any `Connection reset by Peer`.

## What did I learn from this?
Other than double-checking double-underscores, for the umpteenth time I find myself thrilled by issues like this that I know I have the solution at hands reach but still it takes me awhile to be sure of it.

**Comment by floux-digital** [comment]
It was the perfect start for me to find a solution for my case. My stack uses Directus 10.8.3, Postgis 15, both in containers within a docker swarm network.

I tried the solutions you presented, but in my case, what solved the problem was a tip I got on the Directus discord:

```
DB_POOL__MIN: 0
DB_POOL__IDLE_TIMEOUT_MILLIS: 3000
```
